eb_hooks.py

```diff
@@ -353,6 +357,28 @@ def parse_hook_CP2K_remove_deps_for_aarch64(ec, *args, **kwargs):
         raise EasyBuildError("CP2K-specific hook triggered for non-CP2K easyconfig?!")
 
 
+def parse_hook_zen4_module_only(ec, eprefix):
+    """
+    Use --force --module-only if building a foss-2022b-based EasyConfig for Zen4.
+    This toolchain will not be supported on Zen4, so we will generate a modulefile
+    and have it print an LmodError.
+    """
+    ecname, ecversion = ec['name'], ec['version']
+    tcname, tcversion = ec['toolchain']['name'], ec['toolchain']['version']
+    if (
+        (tcname in ['foss', 'gompi'] and tcversion == '2022b') or
+        (tcname in ['GCC', 'GCCcore'] and LooseVersion(tcversion) == LooseVersion('12.2.0')) or
+        (ecname in ['foss', 'gompi'] and ecversion == '2022b') or
+        (ecname in ['GCC', 'GCCcore'] and LooseVersion(ecversion) == LooseVersion('12.2.0'))
+    ):
+        update_build_option('force', 'True')
+        update_build_option('module_only', 'True')
```

Comment:
I think you also need to unset/restore these options again when you are done with the build

Reply:
See how this is done for `EESSI_RPATH_OVERRIDE_ATTR`
* https://github.com/EESSI/software-layer/blob/d61f33d20b7d9d6657b8f4bc45657d4a63a1c243/eb_hooks.py#L173

Reply:
The reason it could be reset is because it is done in a step hook, instead of a parse hook. I'm not sure if that will work here: I'll need to figure out when `--module-only` and `--force` need to be set for this to work, but that might be before the steps even take place. I'll need to dig into the code to be sure...

Reply:
Just as a note, the _current_ `eb_hooks.py` works fine with an EasyStack:

```
$ cat /tmp/casparl/test_easystack.yml
easyconfigs:
  - BLAST+-2.14.0-gompi-2022b.eb
  - gnuplot-5.4.8-GCCcore-12.3.0.eb:
      options:
         rebuild: True
  - gnuplot-5.4.6-GCCcore-12.2.0.eb
 ```

I tested this, and `gnuplot-5.4.8-GCCcore-12.3.0.eb` is correctly build (not just with `--module-only`). I'm pretty sure the environment gets reset in between loop iterations over the items in the EasyStack file. Something like 

```
eb --trace --debug --hooks $HOME/EESSI/software-layer/eb_hooks.py BLAST+-2.14.0-gompi-2022b.eb gnuplot-5.4.8-GCCcore-12.3.0.eb --rebuild --robot
```
indeed fails, as the build options are _not_ reset between the BLAS and gnuplot builds. I.e. in this case `gnuplot-5.4.8-GCCcore-12.3.0.eb` is also built with `--module-only`.

Not very nice, admittedly. Shouldn't be an issue in _our particular scenario_, but I'll check if I can somehow set this early enough in step hooks so that I can reset it at the end.

Reply:
Yeah, this sounds like something that worked differently in an older EasyBuild release but the behaviour has since been improved (I see that happened in https://github.com/easybuilders/easybuild-framework/pull/4213).

Reply:
Tried setting it in the `pre_fetch_hook`, which is the _first_ step hook. It kind-of works. This https://github.com/easybuilders/easybuild-framework/blob/5661bf516341ffc8fcd1518914fbfddbb87bb497/easybuild/framework/easyblock.py#L3915 is where the build option is checked to decide whether to skip a step. By the time the `pre_fetch_hook` is run, that logic has already ran and decided _not_ to skip the fetch step. _But_: it will skip all of the subsequent steps.

The question is whether hooks are still called on steps that are skipped. If not, I'd have to revert the option in the last step run for a `module-only` build _with_ `force`.

Reply:
Ok, setting in `pre_fetch_hook` and unsetting in `post_module_hook` does what it should so. The exception is still that the `fetch` step will _not_ be skipped for `GCCcore-12.2.0`-based stuff on `zen4`, but I can live with that. By resetting the options in a `post_module_hook`, something like:

```
eb --trace --debug --hooks $HOME/EESSI/software-layer/eb_hooks.py BLAST+-2.14.0-gompi-2022b.eb gnuplot-5.4.4-GCCcore-11.3.0.eb gnuplot-5.4.6-GCCcore-12.2.0.eb --rebuild --robot
```

Works as intended. I.e. all the GCCcore-12.2.0 stuff will run with `--module-only --force`, while the `GCCcore-11.3.0` based stuff will just be build as normal.

I'll polish the hooks before making a new commit.

Reply:
Great, I just realised that this would indeed be necessary as the hooks are used by `EESSI-extend` and there is no guarantee people would use easystack files there

Reply:
Yep, I realized the same. Although it would only go wrong if people would build GCCcore-12.2.0 stuff on zen4. But still. Better to do it right :)